use std::vector with CHECK_STORAGE_AGAINST_STANDARD

carmichaelong opened the pull request:
This PR starts to address #1194 by accomplishing two things:
1) CHECK_STORAGE_AGAINST_STANDARD now expects tolerances to be passed in by `std::vector` (instead of `OpenSim::Array`)
2) Changes `Storage::compareWithStandard()` to use `std::vector` internally.

chrisdembia commented:
LGTM. +1 for indenting all those lines of code!

chrisdembia commented:
I think this PR needs one more reviewer.

tkuchida commented on Applications/CMC/test/testCMCArm26_Millard.cpp:
```diff
@@ -45,7 +45,7 @@ void testCMCArm26() {
     cmc.getModel().formStateStorage(temp, *standard);
 
-    Array<double> rms_tols(0.02, 2*2+2*6); // activations within 2%, angles within .6 degrees
+    std::vector<double> rms_tols(2*2+2*6, 0.02); // activations within 2%, angles within .6 degrees
```
I know you didn't put it there, but the trailing comment is obsolete and should be deleted.

klshrinidhi commented on OpenSim/Auxiliary/auxiliaryTestFunctions.h:
```diff
@@ -79,16 +79,16 @@ inline void ASSERT(bool cond,
                                     std::string testFile, 
                                     int testFileLine, 
                                     std::string errorMessage)
 {
```
I think the parameters need to be received as `const T&`. I understand this PR is not introducing the problem but good to fix it if possible.

carmichaelong replied:
I'm not sure if this should be templatized. In the end, `tolerances` is compared to `comparisons`, which is necessarily a vector of `double` because of [this line](https://github.com/opensim-org/opensim-core/blob/4c416d4d000e4cfcbbe59b64f40c6f4140359076/OpenSim/Common/Storage.cpp#L3338). I could easily be missing something though?

chrisdembia replied:
@klshrinidhi is not suggesting to make this templated. He is just using `T` as a placeholder for `std::vector<double>`, `std::string`, etc.

chrisdembia replied:
@klshrinidhi is saying: pass by const reference instead of pass by value.

carmichaelong replied:
That makes way more sense. Thanks!

tkuchida commented on Applications/CMC/test/testCMCArm26_Thelen.cpp:
```diff
@@ -41,7 +41,7 @@ void testCMCArm26() {
     cmc.getModel().formStateStorage(temp, *standard);
 
-    Array<double> rms_tols(0.02, 2*2+2*6); // activations within 2%, angles within .6 degrees
+    std::vector<double> rms_tols(2*2+2*6, 0.02); // activations within 2%, angles within .6 degrees
```
Please delete trailing comment or update `.6` to `1.5`.

klshrinidhi commented on OpenSim/Common/Storage.cpp:
```diff
@@ -3328,23 +3328,19 @@ double Storage::compareColumnRMS(Storage& aOtherStorage, const std::string& aCol
  */
-void Storage::compareWithStandard(Storage& standard, Array<string> &columnsUsed, Array<double> &comparisons)
+void Storage::compareWithStandard(Storage& standard, std::vector<string> &columnsUsed, std::vector<double> &comparisons)
 {
```
Same comment here. Receive arguments as `const T&`.

klshrinidhi commented on OpenSim/Common/Storage.cpp:
```diff
@@ -3328,23 +3328,19 @@ double Storage::compareColumnRMS(Storage& aOtherStorage, const std::string& aCol
-    comparisons.ensureCapacity(maxColumns);
+    columnsUsed.reserve(maxColumns);
+    comparisons.reserve(maxColumns);
 
```
No need to reserve space. This kind of thing is only helps with huge sized arrays of the order of 1000s or elements.

klshrinidhi commented on OpenSim/Common/Storage.cpp:
```diff
@@ -3328,23 +3328,19 @@ double Storage::compareColumnRMS(Storage& aOtherStorage, const std::string& aCol
+            comparisons.push_back(comparison);
+            columnsUsed.push_back(_columnLabels[i]);
         }
     }
```
This loop could use iterators instead.

klshrinidhi commented on OpenSim/Common/Storage.h:
```diff
@@ -303,7 +303,7 @@ OpenSim_DECLARE_CONCRETE_OBJECT(Storage, StorageInterface);
     //void checkAgainstStandard(Storage standard, Array<double> &tolerances, std::string testFile = "", int testFileLine = -1, std::string errorMessage = "Exception");
-    void compareWithStandard(Storage& standard, Array<std::string> &columnsUsed, Array<double> &comparisons);
+    void compareWithStandard(Storage& standard, std::vector<std::string> &columnsUsed, std::vector<double> &comparisons);
     /** Force column labels for a Storage object to become unique. This is done
```
Could receive arguments as `const T&`.

klshrinidhi commented:
:+1:

tkuchida commented on Applications/CMC/test/testCMCWithControlConstraintsGait2354.cpp:
```diff
@@ -66,26 +66,28 @@ void testGait2354() {
         rms_tols2[i] = 0.75; // velocities
     }
-    for (int i = 46; i < rms_tols2.getSize(); ++i){
+    for (int i = 46; i < rms_tols2.size(); ++i){
```
`size()` returns an unsigned so I believe some compilers will give a warning when attempting to compare `i` with `rms_tols2.size()`. Not a big deal (we have lots of warnings), but perhaps `i < (int)rms_tols2.size()`?

carmichaelong replied:
You're right, this should lead to a warning. My instinct would be to change it to `for (size_t i = 46; i < rms_tols2.size(); ++i)` to avoid casting. Thoughts on which is better?

tkuchida replied:
> Thoughts on which is better?

I would make `i` a `size_t` in my own code, but OpenSim tends not to do this (in fact, there are methods like `numCoordinates()` that return ints when they should really return unsigned ints). There are reasons to avoid unsigned ints (see, e.g., #1155 and #1198), but in this particular case I don't see any problem with using `size_t`.

tkuchida commented on OpenSim/Common/Storage.cpp:
```diff
@@ -3328,23 +3328,19 @@ double Storage::compareColumnRMS(Storage& aOtherStorage, const std::string& aCol
-            comparisons[columns] = comparison;
-            columnsUsed[columns++] = _columnLabels[i];
+            comparisons.push_back(comparison);
+            columnsUsed.push_back(_columnLabels[i]);
```
Much clearer! :eyeglasses:

tkuchida commented:
`</review>` :+1: Minor suggestions only; can be ignored.

tkuchida commented:
Oops, sorry I didn't notice @klshrinidhi was already reviewing :see_no_evil:

carmichaelong commented:
Addressed comments in the last commit and tests are passing. Re: iterator for the for loop: the member variable looped over is an OpenSim::Array and not a std::vector. I think changing that member variable is out of scope for this PR.

tkuchida commented on OpenSim/Auxiliary/auxiliaryTestFunctions.h:
```diff
@@ -77,18 +77,18 @@ inline void ASSERT(bool cond,
+void CHECK_STORAGE_AGAINST_STANDARD(const OpenSim::Storage& result, 
+                                    const OpenSim::Storage& standard, 
+                                    const std::vector<double>& tolerances, 
+                                    const std::string testFile, 
```
I think the two `const std::string`s should be `const std::string&`.

carmichaelong replied:
Ah, yes. Thanks, I got a little distracted by the `const` and forgot the references there. This also made me look at the `int` line, and apparently there's some thought that passing primitives by reference may not speed things up (one of many sources [here](http://stackoverflow.com/questions/3009543/passing-integers-as-constant-references-versus-copying))

tkuchida replied:
> there's some thought that passing primitives by reference may not speed things up

:+1: This thought jibes with the convention used in Simbody and OpenSim.

tkuchida commented:
`</re-review>` One minor comment; can be merged without it IMO. :+1:

carmichaelong commented:
Thanks @tkuchida @klshrinidhi @chrisdembia!
